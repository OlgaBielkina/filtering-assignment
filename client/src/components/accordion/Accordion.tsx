import React, { useState } from 'react';

interface Props {
    children: React.ReactNode;
}

export interface IAccordionContext {
    expanded: boolean;
    handleExpanded?: () => void;
}

export const AccordionContext = React.createContext<IAccordionContext>({
    expanded: false,
    handleExpanded: undefined,
});

export const Accordion: React.FC<Props> = ({ children }) => {
    const [expanded, setExpanded] = useState(false);
    const handleExpanded = () => setExpanded(!expanded);
    const data = { expanded, handleExpanded };

    return <AccordionContext.Provider value={data}>{children}</AccordionContext.Provider>;
};
