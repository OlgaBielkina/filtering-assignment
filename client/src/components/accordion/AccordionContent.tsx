import React, { useContext } from 'react';
import { AccordionContext } from './Accordion';
import styles from './AccordionContent.module.scss';

interface Props {
    children: React.ReactNode;
}

export const AccordionContent: React.FC<Props> = ({ children }) => {
    const context = useContext(AccordionContext);

    return (
        <div data-testid="accordionContent" className={context.expanded ? '' : styles.hidden}>
            {children}
        </div>
    );
};
