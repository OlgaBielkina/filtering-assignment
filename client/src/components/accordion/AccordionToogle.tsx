import React, { useContext } from 'react';
import classnames from 'classnames';
import { AccordionContext } from './Accordion';
import styles from './AccordionToogle.module.scss';

export interface Props {
    children: React.ReactNode;
    collapsible: boolean;
}

export const AccordionToogle: React.FC<Props> = ({ children, collapsible }) => {
    const context = useContext(AccordionContext);
    const handler = (e: React.MouseEvent<HTMLDivElement>) => {
        e.stopPropagation();
        if (collapsible && context.handleExpanded) {
            context.handleExpanded();
        }
    };

    return (
        <div className={styles.root} onClick={handler}>
            {children}
            {collapsible && (
                <span className={classnames('material-icons', styles.icon)}>
                    {context.expanded ? 'expand_less' : 'expand_more'}
                </span>
            )}
        </div>
    );
};
