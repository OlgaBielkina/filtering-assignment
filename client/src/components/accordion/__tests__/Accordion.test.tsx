import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import { AccordionContent } from '../AccordionContent';
import { Accordion } from '../Accordion';
import { AccordionToogle } from '../AccordionToogle';

describe('Accordion', () => {
    it('should render collapsed content', () => {
        const { asFragment, getByText } = render(
            <Accordion>
                <AccordionToogle collapsible>toggle</AccordionToogle>
                <AccordionContent>test text</AccordionContent>
            </Accordion>,
        );

        expect(getByText('test text')).toHaveClass('hidden');
        expect(asFragment()).toMatchSnapshot();
    });

    it('should expand content when user clicks toggle', () => {
        const { getByText } = render(
            <Accordion>
                <AccordionToogle collapsible>toggle</AccordionToogle>
                <AccordionContent>test text</AccordionContent>
            </Accordion>,
        );

        fireEvent.click(getByText('toggle'));

        expect(getByText('test text')).not.toHaveClass('hidden');
    });
});
