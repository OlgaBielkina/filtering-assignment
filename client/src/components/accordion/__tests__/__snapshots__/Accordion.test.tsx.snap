// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion should render collapsed content 1`] = `
<DocumentFragment>
  <div
    class="root"
  >
    toggle
    <span
      class="material-icons icon"
    >
      expand_more
    </span>
  </div>
  <div
    class="hidden"
    data-testid="accordionContent"
  >
    test text
  </div>
</DocumentFragment>
`;
